SliderButton/SliderButtonBehaviorDecorator.js

```diff
@@ -10,43 +12,71 @@ import {findDOMNode} from 'react-dom';
  * @memberof agate/SliderButton
  * @private
  */
-const SliderButtonBehaviorDecorator = (Wrapped) => {
-	// eslint-disable-next-line no-shadow
-	function SliderButtonBehaviorDecorator ({onChange, ...rest}) {
-		const {children} = rest;
-		const [valueText, setValueText] = React.useState(children ? children[0] : null);
-		const ref = React.useRef();
-
-		function handleChange ({value}) {
-			onChange();
-			setValueText(children[value]);
+const SliderButtonBehaviorDecorator = hoc((config, Wrapped) => {
+	return class extends React.Component {
+		static displayName = 'SliderButtonBehaviorDecorator';
+
+		static propTypes = /** @lends agate/ArcSlider.ArcSliderBehaviorDecorator.prototype */ {
+			/**
+			 * Items displayed with SliderButton.
+			 *
+			 * @type {Function}
+			 * @private
+			 */
+			children: PropTypes.node,
+
+			/**
+			 * Called when value is changed.
+			 *
+			 * @type {Function}
+			 * @private
+			 */
+			onChange: PropTypes.func
+		};
+
+		constructor (props) {
+			super(props);
+
+			this.ref = React.createRef();
+
+			this.state = {
+				valueText: props.children ? props.children[0] : null
+			};
 		}
 
-		function handleDragStart () {
+		handleChange = ({value}) => {
+			this.setState(
+				() => ({valueText: this.props.children[value]}),
+				() => {
+					forward('onChange', {
+						type: 'onChange',
+						value
+					}, this.props);
+				}
+			);
+		};
+
+		handleDragStart = () => {
 			// on platforms with a touchscreen, we want to focus slider when dragging begins
 			if (platform.touchscreen) {
-				findDOMNode(ref.current).focus(); // eslint-disable-line react/no-find-dom-node
+				findDOMNode(ref.current).focus(); // eslint-disable-line
 			}
-		}
+		};
 
-		return (
-			<Wrapped
-				aria-valuetext={valueText}
-				ref={ref}
-				role="slider"
-				{...rest}
-				onChange={handleChange}
-				onDragStart={handleDragStart}
-			/>
-		);
-	}
-
-	SliderButtonBehaviorDecorator.propTypes = {
-		onChange: PropTypes.func
+		render () {
+			return (
+				<Wrapped
+					{...this.props}
+					aria-valuetext={this.state.valueText}
+					onChange={this.handleChange}
+					onDragStart={this.handleDragStart}
+					ref={this.ref}
+					role="slider"
```

Comment:
We intended to add `aria-valuetext` and `role` before `{...rest}` because app developers could override them by passing props. It's the common way with A11y props. It breaks our coding rule, but it is easy to provide the way to override component default A11y features.

Reply:
Sorry, I don't quite understand your point here. Do you mean I should pass `{...this.props}` at the bottom like this?

```suggestion
					return (
				<Wrapped
					aria-valuetext={this.state.valueText}
					onChange={this.handleChange}
					onDragStart={this.handleDragStart}
					ref={this.ref}
					role="slider"
					{...this.props}
				/>
			);
```

Reply:
My expectation is the following one.
Only a11y related props are defined before `{...this.props}`

```
		render () {
			return (
				<Wrapped
					aria-valuetext={this.state.valueText}
					role="slider"
					{...this.props}
					onChange={this.handleChange}
					onDragStart={this.handleDragStart}
					ref={this.ref}
				/>
			);
		}
```

Reply:
I see. Thank you!